File: BaekjoonAlgorithm/BA_1325/BJAlgo_1325.py
from sys import stdin
from collections import deque


def bfs(v):
    q = deque([v])
    visited = [False] * (n + 1)
    visited[v] = True
    count = 1
    while q:
        v = q.popleft()
        for e in adj[v]:
            if not visited[e]:
                q.append(e)
                visited[e] = True
                count += 1

    return count


n, m = map(int, stdin.readline().split())
adj = [[] for _ in range(n + 1)]

for _ in range(m):
    a, b = map(int, stdin.readline().split())
    adj[b].append(a)

result = []
max_value = -1

for i in range(1, n + 1):
    c = bfs(i)
    if c > max_value:
        max_value = c
        result = [i]
    elif c == max_value:
        result.append(i)

for r in result:
    print(r, end=" ")

File: BaekjoonAlgorithm/202012/BA_1325/BJAlgo_1325.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class BJAlgo_1325 {
    private static List<ArrayList<Integer>> adj = new ArrayList<>();

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int[] nm = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        for (int i = 0; i < nm[0] + 1; i++) {
            adj.add(i, new ArrayList<>());
        }

        for (int i = 0; i < nm[1]; i++) {
            int[] ab = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
            adj.get(ab[1]).add(ab[0]);
        }

        ArrayList<Integer> result = new ArrayList<>();
        int max_value = -1;

        for (int i = 1; i < adj.size(); i++) {
            int c = bfs(i);

            if (c > max_value) {
                result.clear();
                result.add(i);
                max_value = c;
            } else if (c == max_value) {
                result.add(i);
            }
        }

        for (int r : result) {
            sb.append(r).append(" ");
        }

        System.out.println(sb);
    }

    private static int bfs(int v) {
        Deque<Integer> deque = new ArrayDeque<>();
        deque.add(v);

        boolean[] visited = new boolean[adj.size() + 1];
        int count = 1;

        while (!deque.isEmpty()) {
            int n = deque.pollFirst();
            for (int e : adj.get(n)) {
                if (!visited[e]) {
                    deque.add(e);
                    visited[e] = true;
                    count += 1;
                }
            }
        }

        return count;
    }
}
